pkg/apis/pipeline/v1alpha1/taskrun_types.go

```diff
@@ -68,6 +68,8 @@ const (
 	// TaskRunSpecStatusCancelled indicates that the user wants to cancel the task,
 	// if not already cancelled or terminated
 	TaskRunSpecStatusCancelled = v1beta1.TaskRunSpecStatusCancelled
+	// TaskRunSpecStatusSkipped indicates that the task should be skipped and not executed
```

Comment:
With these changes, `TaskRun` is created for a task which depends on a certain state of a parent but parent ends up with different state, for example, `TaskB` depends on failure of `TaskA` but `TaskA` finishes with success. In this scenario, a taskrun is created for `TaskB` with this state, more specifically `reason`.  This state is very crucial (1) to determine the status of PipelineRun, Completed with a skipped task (2) tasks which are dependent on `TaskB` are also marked with the same state instead of failing the whole pipeline

Note these two examples and how `runOn` handles them:

https://github.com/tektoncd/pipeline/blob/64d326f29cee582d7f5a75ae6e9e73fe691cc289/docs/pipelines.md#state-of-a-parent-task-specified-in-runon-section-conflicts-with-the-actual-state

https://github.com/tektoncd/pipeline/blob/64d326f29cee582d7f5a75ae6e9e73fe691cc289/docs/pipelines.md#task-with-multiple-dependencies-based-on-different-states

Reply:
OK, I see, thanks for the clarification!

I think this calls for a bit of extra dicussion, since the proposed behaviour diverges from what we do today:
- if a task fails, the pipeline is failed. Any task that was running until that point finishes, and no new task is scheduled. Tasks that are not executed have no taskrun associated and are not tracked in the pipelinerun status either
- if a task if not executed because of a `runAfter` condition that is not met, the taskrun is not created and not tracked in the pipelinerun status
- if a task does not pass a condition, it is not executed, so no taskrun is created for it. However it is tracked in the pipelinerun status (see example below), because the conditions are part of it. tasks that depend on a task skipped because of a condition are not executed and not tracked in the pipelinerun status

```
      status:
        conditions:
        - lastTransitionTime: "2020-03-12T21:02:42Z"
          message: ConditionChecks failed for Task tekton-random-check-xsz9w-check-reset-dr9hr
            in PipelineRun tekton-random-check-xsz9w
          reason: ConditionCheckFailed
          status: "False"
          type: Succeeded
        podName: ""
```

For `runOn`, to be consistent with today's behaviour for `runAfter`, we should not create a taskrun and not track it in the pipelinerun status. 
If we wanted to be consisted with the conditions instead, we could track the evaluation of the runOn in the pipelinerun status, but not create a taskrun.

This said I'm not necessarily advocating for our current approach. It is straight-forward and fairly readable, but it has a downside: anyone (CLI, Dashboard, human user) trying to understand what happened in the pipeline, need to consult the pipeline definition, infer the DAG and compare with the known states to fully understand what happened. For fairly simple pipelines this is easy to do, but it may not always be so obvious for more complex ones, especially in case of humans looking at API output directly. 

One thing that we may consider doing in future is to track all taskruns in the pipelinerun status; if we do so we should for all cases, and I think the change should not happen in this PR.

Something I would not do in any case is to actually submit new taskruns resources to the API only to mark them as skipped. That information should be inferred from the parent object, the pipelinerun.

@dibyom @vdemeester @sbwsg @bobcatfish thoughts?

Reply:
Sure, it would be great to discuss this further and get everyone's input 👍 

This is certainly different than the way we handle conditional failures today. The biggest difference with `runOn` is pipeline continues even after a conflict is discovered.

Today, we stop pipelinerun as soon as a task has failed or condition resulted in failure. In case when a task fails, `pipelinerun` would have the `taskrun` of the failed task with the failure reason. 

```
tkn pr describe 01-pipelinerun-without-runon

💌 Message

TaskRun 01-pipelinerun-without-runon-task1-ln8zz has failed ("step-fail" exited with code 1 (image: "docker-pullable://ubuntu@sha256:04d48df82c938587820d7b6006f5071dbbffceb7ca01d2814f81857c631d44df"); for logs run: kubectl -n default logs 01-pipelinerun-without-runon-task1-ln8zz-pod-pjgmg -c step-fail)

🗂  Taskruns

 NAME                                         TASK NAME   STARTED          DURATION    STATUS
 ∙ 01-pipelinerun-without-runon-task1-ln8zz   task1       13 seconds ago   8 seconds   Failed
```

Now, its possible to retrieve the `taskrun`:

```
tkn tr describe 01-pipelinerun-without-runon-task1-ln8zz

Message

"step-fail" exited with code 1 (image: "docker-pullable://ubuntu@sha256:04d48df82c938587820d7b6006f5071dbbffceb7ca01d2814f81857c631d44df"); for logs run: kubectl -n default logs 01-pipelinerun-without-runon-task1-ln8zz-pod-pjgmg -c step-fail

 NAME     STATUS
 ∙ fail   Error
```

Whereas for a task with a condition and when that condition fails, `pipelinerun` contains `taskrun` for the task:

```
tkn pr describe pipelinerun-condition-failures
🗂  Taskruns

 NAME                                           TASK NAME   STARTED   DURATION   STATUS
 ∙ pipelinerun-condition-failures-task1-j46vj   task1       ---       ---        Failed(ConditionCheckFailed)
```

And can not retrieve that `taskrun`:

```
tkn tr describe pipelinerun-condition-failures-task1-j46vj
Error: failed to find taskrun "pipelinerun-condition-failures-task1-j46vj"
``` 
But at least `taskrun` for the condition exists and can be retrieved:

```
tkn tr list
NAME                                                              STARTED          DURATION    STATUS
pipelinerun-condition-failures-task1-j46vj-false-conditio-gkllw   8 minutes ago    7 seconds   Failed
```

In case of `runOn`, if a `taskrun` is listed in the `pipelinerun` as conflicted, providing an ability to retrieve that `taskrun` would be helpful. But we can certainly drop it out of this PR if needed. 

```
tkn pr describe pipelinerun-runon-conflicted
🌡️  Status

STARTED          DURATION   STATUS
17 minutes ago   1 minute   Succeeded(Completed)

🗂  Taskruns

 NAME                                                            TASK NAME   STARTED          DURATION     STATUS
 ∙ pipelinerun-runon-conflicted-task2-s942v   task2       16 minutes ago   23 seconds   Failed
 ∙ pipelinerun-runon-conflicted-task3-8mhh9   task3       16 minutes ago   ---          Failed(SkippedAsStateConflicted)
 ∙ pipelinerun-runon-conflicted-task1-trxwl   task1       17 minutes ago   22 seconds   Succeeded
```

```
 tkn tr describe pipelinerun-runon-conflicted-task3-8mhh9
🌡️  Status

STARTED          DURATION    STATUS
17 minutes ago   ---         Failed(SkippedAsStateConflicted)

Message

TaskRun "pipelinerun-runon-conflicted-task3-8mhh9" was skipped
```

Reply:
> Today, we stop pipelinerun as soon as a task has failed or condition resulted in failure. In case when a task fails, `pipelinerun` would have the `taskrun` of the failed task with the failure reason.

What do you mean by "condition resulted in failure"?
If the check in a condition is false, so the condition fails, the associated task is not executed, however the execution of the pipelinerun continues.

Reply:
> And can not retrieve that `taskrun`:
> 
> ```
> tkn tr describe pipelinerun-condition-failures-task1-j46vj
> Error: failed to find taskrun "pipelinerun-condition-failures-task1-j46vj"
> ```

Personally I think this correct, the task was not executed at all and so it was not created.
The information about what happened is available in the pipelinerun. 
If we create a taskrun, even as a failed one, it will trigger the taskrun controller, which will emit events and do some work.

Reply:
> This said, I'm not necessarily advocating for our current approach. It is straight-forward and fairly readable, but it has a downside: anyone (CLI, Dashboard, human user) trying to understand what happened in the pipeline, need to consult the pipeline definition, infer the DAG and compare with the known states to fully understand what happened. For fairly simple pipelines this is easy to do, but it may not always be so obvious for more complex ones, especially in case of humans looking at API output directly.

@afrittoli I think we need to capture this in an issue and work towards having a `status` on `PipelineRun` that is *self descriptive* in some ways (aka making it easier for cli, dashboard, .) to get information without having to query the `Pipeline` spec. 

cc @bobcatfish

Reply:
Big +1 to tracking the entire state using the pipelinerun status instead of looking up taskruns